<a href="https://colab.research.google.com/github/brendankilduff/assignment2/blob/main/Kilduff_Brendan_Assignment_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [7]:
import numpy as np # Import the numpy package into your workspace
import pandas as pd  # Import the pandas package into your workspace
df = pd.read_csv('/content/airbnb_hw.csv',low_memory=False)
print(df.shape, '\n') # List the dimensions of df
print(df.dtypes, '\n') # The types of the variables; `object` is a bad sign
print(df.columns[1:5], '\n') # First five column names

# Clean the 'Price' variable
df['Price'] = df['Price'].str.replace(',', '', regex=True)  # Remove commas
df['Price'] = df['Price'].apply(pd.to_numeric, errors='coerce')  # Convert to floats and handle missing values
df['Price'] = df['Price'].replace(' ',np.nan) # Notice the column replacement

# Print the count
print("Number of NaN values in 'Price' column:", nan_count)



(30478, 13) 

Host Id                         int64
Host Since                     object
Name                           object
Neighbourhood                  object
Property Type                  object
Review Scores Rating (bin)    float64
Room Type                      object
Zipcode                       float64
Beds                          float64
Number of Records               int64
Number Of Reviews               int64
Price                          object
Review Scores Rating          float64
dtype: object 

Index(['Host Since', 'Name', 'Neighbourhood ', 'Property Type'], dtype='object') 

Number of NaN values in 'Price' column: 274


For cleaning the Price variable, I made sure to delete commas, turn the values into floats, and make sure blank values go into the nan count.

In [ ]:

df = pd.read_csv('/content/sharks.csv',low_memory=False)
print(df.shape, '\n') # List the dimensions of df
print(df.dtypes, '\n') # The types of the variables; `object` is a bad sign
print(df.columns[1:5], '\n') # First five column names
print(df['Type'].unique(), '\n') # A Categorical Example
df['Type'] = df['Type'].replace(' ',np.nan) # Notice the column replacement

replacement_mapping = {
    'Boating': 'Watercraft',
    'Boatomg': 'Watercraft',
    'Boat' : 'Watercraft',
    'Unverified': 'Questionable',
    'Under investigation' : 'Questionable',
    'Unconfirmed' : 'Questionable',
    }

# Use the replace method to perform the replacement
df['Type'] = df['Type'].replace(replacement_mapping)
print(df['Type'].value_counts(), '\n')

nan_count = df['Type'].isna().sum()

# Print the count
print("Number of NaN values in 'Type' column:", nan_count)


(6462, 257) 

index             int64
Case Number      object
Date             object
Year            float64
Type             object
                 ...   
Unnamed: 251    float64
Unnamed: 252    float64
Unnamed: 253    float64
Unnamed: 254    float64
Unnamed: 255    float64
Length: 257, dtype: object 

Index(['Case Number', 'Date', 'Year', 'Type'], dtype='object') 

['Unprovoked' 'Provoked' 'Questionable' 'Watercraft' 'Unconfirmed'
 'Unverified' 'Invalid' 'Under investigation' 'Boating' 'Sea Disaster' nan
 'Boat' 'Boatomg'] 

Number of NaN values in 'Type' column: 5
Unprovoked      4716
Provoked         593
Invalid          552
Watercraft       344
Sea Disaster     239
Questionable      13
Name: Type, dtype: int64 

Number of NaN values in 'Type' column: 5


I cleaned the sharks variable by replacing blank data with NaN and then combining some of the shark attack types together as they overlap in definition.

In [12]:
df = pd.read_csv('/content/Pretrialdata2017.csv',low_memory=False)
print(df.shape, '\n') # List the dimensions of df
print(df.dtypes, '\n') # The types of the variables; `object` is a bad sign
print(df['WhetherDefendantWasReleasedPretrial'].unique(), '\n') # A Categorical Example
df['WhetherDefendantWasReleasedPretrial'] = df['WhetherDefendantWasReleasedPretrial'].replace( 9 ,np.nan)
print(df['WhetherDefendantWasReleasedPretrial'].value_counts(), '\n')

nan_count = df['WhetherDefendantWasReleasedPretrial'].isna().sum()

# Print the count
print("Number of NaN values in 'WhetherDefendantWasReleasedPretrial' column:", nan_count)



(22986, 709) 

InternalStudyID                                                 object
REQ_REC#                                                        object
Defendant_Sex                                                   object
Defendant_Race                                                  object
Defendant_BirthYear                                             object
                                                                 ...  
NewFelonySexualAssaultArrest_Disposition                        object
Intertnalindicator_ReasonforExcludingFromFollowUpAnalysis        int64
CriminalHistoryRecordsReturnedorCMSRecordsFoundforIndividual    object
DispRecordFoundforChargesinOct2017Contact_Atleast1dispfound      int64
CrimeCommission2021ReportClassificationofDefendants             object
Length: 709, dtype: object 

[9 0 1] 

1.0    19154
0.0     3801
Name: WhetherDefendantWasReleasedPretrial, dtype: int64 

Number of NaN values in 'WhetherDefendantWasReleasedPretrial' column: 31


I cleaned the variable 'WhetherDefendantWasReleasedPretrial' by replacing any 9 with a NaN value because the codebook implies that these values are missing.

In [15]:
df = pd.read_csv('/content/Pretrialdata2017.csv',low_memory=False)
print(df.shape, '\n') # List the dimensions of df
print(df.dtypes, '\n') # The types of the variables; `object` is a bad sign
print(df['ImposedSentenceAllChargeInContactEvent'].describe())
df['ImposedSentenceAllChargeInContactEvent'] = df['ImposedSentenceAllChargeInContactEvent'].apply(pd.to_numeric, errors='coerce')
df['ImposedSentenceAllChargeInContactEvent'] = df['ImposedSentenceAllChargeInContactEvent'].replace(' ',np.nan) # Notice the column replacement
print(df['ImposedSentenceAllChargeInContactEvent'].value_counts(), '\n')

print(df['SentenceTypeAllChargesAtConvictionInContactEvent'].unique(), '\n') # A Categorical Example
print(df['SentenceTypeAllChargesAtConvictionInContactEvent'].value_counts(), '\n')

df.loc[df['SentenceTypeAllChargesAtConvictionInContactEvent'] == 4, 'ImposedSentenceAllChargeInContactEvent'] = 0
nan_count = df['ImposedSentenceAllChargeInContactEvent'].isna().sum()
print("Number of NaN values in 'Type' column:", nan_count)

#all missing data is either a 4 or 9. If it is a 9, it is missing, while a 4 turns into 0.
print(df['ImposedSentenceAllChargeInContactEvent'].value_counts(), '\n')





(22986, 709) 

InternalStudyID                                                 object
REQ_REC#                                                        object
Defendant_Sex                                                   object
Defendant_Race                                                  object
Defendant_BirthYear                                             object
                                                                 ...  
NewFelonySexualAssaultArrest_Disposition                        object
Intertnalindicator_ReasonforExcludingFromFollowUpAnalysis        int64
CriminalHistoryRecordsReturnedorCMSRecordsFoundforIndividual    object
DispRecordFoundforChargesinOct2017Contact_Atleast1dispfound      int64
CrimeCommission2021ReportClassificationofDefendants             object
Length: 709, dtype: object 

count     22986
unique      484
top            
freq       9053
Name: ImposedSentenceAllChargeInContactEvent, dtype: object
0.000000     4953
12.000000    1404
0.985626     10

To clean this data, I checked the relationship between imposed sentence and sentence type. The only missing values should be those that correlate with 9 on the senetence type variable. If they correlated with 4, there was no incarceration, so they turned to 0. I made sure to turn missing values into NaN and turn the values into floats. I then turned all values that had a 4 in sentence type equal to 0. The remaining 274 observations that were equal to 9 are the only missing values in the imposed sentence data.  